<a href="https://colab.research.google.com/github/GuilleLondero/challenge_one_DS_telecomX/blob/main/ChallengerTelecomX.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**📌 Extracción**

In [2]:
# 1. Importar las bibliotecas necesarias
import pandas as pd

# 2. URL del archivo JSON estático en GitHub
url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json"

# 3. Leo el contenido JSON desde la URL y convierto los datos a un DataFrame de Pandas
df = pd.read_json(url)

# **Comprensión del DataSet**

In [3]:
# 1. Mostrar información básica del DataFrame
df.shape

(7267, 6)

In [4]:
# 2. Mostrar columnas
df.columns

Index(['customerID', 'Churn', 'customer', 'phone', 'internet', 'account'], dtype='object')

In [5]:
# 3. Tipos de datos de cada columna
df.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [6]:
# 4. Mostrar las primeras 5 filas
df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [7]:
# 5. Información detallada del DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [8]:
# 6. Explorar la columna target (Churn) - Variable objetivo
df['Churn'].value_counts() # Distribución de clientes que se fueron vs se quedaron

Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

In [16]:
# 6.1 Ver todos los valores únicos en Churn
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

Tenemos 224 registros "misteriosos":
Los valores únicos son: ['No', 'Yes', '']
¡Aquí tenemso un problema! Hay 224 registros con valores vacíos ('') en la columna Churn.

Lo que significa:

🟢 'No' = 5174 → Clientes que se quedaron

🔴 'Yes' = 1869 → Clientes que se fueron

⚠️ '' (vacío) = 224 → Registros sin información de Churn

Por eso en el value_counts() aparecía:

No: 5174

Yes: 1869

224 (esta línea extra eran los strings vacíos)

Total verificado:
5174 + 1869 + 224 = 7267 ✅ (coincide con el total del dataset)

¿Qué hacer con estos 224 registros?

In [23]:
# 7. Porcentaje de distribución de Churn
churn_porcentajes = df['Churn'].value_counts(normalize=True) * 100 # Porcentajes para entender el balance
churn_porcentajes.round(2)

Churn
No     71.20
Yes    25.72
        3.08
Name: proportion, dtype: float64

### Interpretación empresarial del resultado anterior:
🎯 Tasa de Churn = 25.72% es MUY ALTA para telecomunicaciones

📊 Retención = 71.20% podría mejorar significativamente

⚠️ 3.08% sin datos necesita limpieza

¿Qué significa esto?
**De cada 100 clientes, aproximadamente 26 se van cada período.**

Esta tasa de churn es preocupante. ***El objetivo será identificar qué factores llevan a ese 25.72% a irse***

## **🧹 Calidad de los datos**

In [10]:
# 1. Verificar si hay valores nulos en las columnas principales
df.isnull().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [11]:
# 2. Verificar si hay customerIDs duplicados
df['customerID'].duplicated().sum()

np.int64(0)

In [12]:
# 3. Estadísticas descriptivas de las columnas numéricas visibles
df.describe()

customerID Churn                                           customer  \
count         7267  7267                                               7267   
unique        7267     3                                                891   
top     9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
freq             1  5174                                                223   

                                                 phone  \
count                                             7267   
unique                                               3   
top     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
freq                                              3495   

                                                 internet  \
count                                                7267   
unique                                                129   
top     {'InternetService': 'No', 'OnlineSecurity': 'N...   
freq                                                 1581   

                                                  account  
count                                                7267  
unique                                               6931  
top     {'Contract': 'Month-to-month', 'PaperlessBilli...  
freq                                                    6

In [24]:
# 4. Explorar la estructura anidada de las columnas
# Columna 'customer' - Datos demográficos
df['customer'].iloc[0]

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

In [25]:
# 4.1. Columna 'phone' - Servicios telefónicos
df['phone'].iloc[0]

{'PhoneService': 'Yes', 'MultipleLines': 'No'}

In [26]:
# 4.2. Columna 'internet' - Servicios de internet
df['internet'].iloc[0]

{'InternetService': 'DSL',
 'OnlineSecurity': 'No',
 'OnlineBackup': 'Yes',
 'DeviceProtection': 'No',
 'TechSupport': 'Yes',
 'StreamingTV': 'Yes',
 'StreamingMovies': 'No'}

In [27]:
# 4.3. Columna 'account' - Información de cuenta y facturación
df['account'].iloc[0]

{'Contract': 'One year',
 'PaperlessBilling': 'Yes',
 'PaymentMethod': 'Mailed check',
 'Charges': {'Monthly': 65.6, 'Total': '593.3'}}